Add lazy-loading support for user metadata.

spacedmonkey opened the pull request:
Introduce the `wp_lazyload_user_meta` function to queue user metadata for lazy-loading, optimizing user meta retrieval. Update related functions, tests, and metadata lazyloader to support this functionality, ensuring improved performance and consistency.

<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/63021
---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

spacedmonkey committed:
Add lazy-loading support for user metadata.

Introduce the `wp_lazyload_user_meta` function to queue user metadata for lazy-loading, optimizing user meta retrieval. Update related functions, tests, and metadata lazyloader to support this functionality, ensuring improved performance and consistency.

westonruter commented on src/wp-includes/user.php:
```diff
@@ -1284,6 +1284,21 @@ function update_user_meta( $user_id, $meta_key, $meta_value, $prev_value = '' )
+ *
+ * @since 6.8.0
+ *
+ * @param array $user_ids List of user IDs.
```
```suggestion
 * @param int[] $user_ids List of user IDs.
```

westonruter commented on src/wp-includes/user.php:
```diff
@@ -1284,6 +1284,21 @@ function update_user_meta( $user_id, $meta_key, $meta_value, $prev_value = '' )
+/**
+ * Queue user meta for lazy-loading.
+ *
+ * @since 6.8.0
```
```suggestion
 * @since 6.9.0
```

spacedmonkey committed:
Apply suggestions from code review

Co-authored-by: Weston Ruter <westonruter@gmail.com>

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-loading-user-meta

westonruter commented on src/wp-includes/class-wp-user-query.php:
```diff
@@ -884,12 +884,15 @@ public function query() {
 				$result->id = $result->ID;
 			}
 		} elseif ( 'all_with_meta' === $qv['fields'] || 'all' === $qv['fields'] ) {
+			$user_ids = array_map( 'intval', $this->results );
```
Maybe this is overkill, but since `intval( array( 'foo' ) )` results in `1`, I think it would be preferable to filter out anything that isn't a number.

```suggestion
			$user_ids = array_filter(
				array_map(
					static function ( $user_id ): ?int {
						return is_numeric( $user_id ) ? intval( $user_id ) : null;
					},
					$this->results
				)
			);
```

Also, the above return value of the `users_pre_query` filter is checked for whether or not it is `null` but not whether it is then an `array` as expected. So I think adding that check to be sure it is an array would be good.

spacedmonkey replied:
Revert changes to this class as not needed - https://github.com/WordPress/wordpress-develop/pull/8411/commits/bf504853a1c32444f647db480bef2fa42f4829d6

westonruter commented on src/wp-includes/class-wp-user-query.php:
```suggestion
		if ( ! $this->results || ! is_array( $this->results ) ) {
```

spacedmonkey replied:
Revert changes to this class as not needed - https://github.com/WordPress/wordpress-develop/pull/8411/commits/bf504853a1c32444f647db480bef2fa42f4829d6

westonruter commented:
@spacedmonkey  I'm trying to follow [your testing instructions](https://core.trac.wordpress.org/ticket/63021#comment:17), but I'm having trouble.

1. I've activated the Twenty Nineteen theme.
2. I've created two pages: One without an Author block and another with an Author block.
4. I only have Query Monitor active.
5. I'm viewing Page template.

# Trunk

Without Author Block | With the Author Block
--|--
<img width="770" height="543" alt="image" src="https://github.com/user-attachments/assets/65b90306-7319-496c-9ae4-c027b70d2c10" /> | <img width="754" height="543" alt="image" src="https://github.com/user-attachments/assets/738f01bb-fef8-4f3e-b9a9-a3ada0c9e240" />



# PR

Without Author Block | With the Author Block
--|--
<img width="771" height="532" alt="image" src="https://github.com/user-attachments/assets/2d54a6ce-0518-418e-a200-9d7edd441f7e" /> | <img width="779" height="546" alt="image" src="https://github.com/user-attachments/assets/2e060bd1-fbe8-4389-a773-72b5d72025b5" />

👉🏻 In all cases, I see the same count of 21 queries. What am I missing?

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-loading-user-meta

spacedmonkey committed:
Revert changes to user query.

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-loading-user-meta

spacedmonkey commented:
@westonruter Sorry for the confused but to see the benefit, [this](https://github.com/WordPress/wordpress-develop/commit/2de7ed3bbc7263dd1f82ab43bd3a49cb46e6ae99) must first be committed.

Now it is committed and this PR rebased, I am seeing the benefit. 
Trunk | PR
--|--
<img width="1512" height="503" alt="Screenshot 2025-10-08 at 21 20 18" src="https://github.com/user-attachments/assets/9651f17f-eb4f-43c1-b833-a02e8d0de56e" />73 queries | <img width="1512" height="588" alt="Screenshot 2025-10-08 at 21 20 05" src="https://github.com/user-attachments/assets/cb922faf-3287-4137-99dc-0bbd617b10b5" />72 queries
<img width="1512" height="510" alt="Screenshot 2025-10-08 at 21 23 57" src="https://github.com/user-attachments/assets/114a2a61-3953-43a1-882a-dcd2be38f6a5" /> 67 queries| <img width="1512" height="516" alt="Screenshot 2025-10-08 at 21 22 33" src="https://github.com/user-attachments/assets/7c40bfbd-bb0f-437f-9c44-438eaa2089f9" /> 66 queries

I am seeing one less query on a number of different types of page load. 

The query I am no longer seeing this one. 
<img width="1324" height="463" alt="Screenshot 2025-10-08 at 21 28 25" src="https://github.com/user-attachments/assets/725e9cfc-dc8c-44b8-a243-fb092e50bfbb" />

This means 28 less rows of user meta loaded. For sites with lots of user meta or multisite, this could be a massive win.

spacedmonkey committed:
Update tests to reflect lazy-loading behavior for user metadata.

spacedmonkey committed:
Add test coverage for forced loading of author metadata

Enhance tests to verify behavior when author metadata is explicitly loaded using `get_the_author_meta`, ensuring user IDs are properly primed.

spacedmonkey committed:
Expand test coverage for `wp_lazyload_user_meta`

Add new tests to handle various scenarios, including users not in the lazy load queue. Update existing tests to reduce created user count and include cache clearing, ensuring more robust and efficient validation of lazy-loading behavior.

westonruter reviewed (approved):
I was finally able to replicate the reduction in queries on the single template, but only after I set the First Name, Last Name, and Display Name for the author of the post I was viewing:

```diff
--- before.sql	2025-10-09 18:21:53
+++ after.sql	2025-10-09 18:20:55
@@ -11,7 +11,6 @@
 SELECT wp_posts.* FROM wp_posts WHERE ID IN (518,288);
 SELECT post_id, meta_key, meta_value FROM wp_postmeta WHERE post_id IN (518,288) ORDER BY meta_id ASC;
 SELECT DISTINCT t.term_id, tr.object_id FROM wp_terms AS t INNER JOIN wp_term_taxonomy AS tt ON t.term_id = tt.term_id INNER JOIN wp_term_relationships AS tr ON tr.term_taxonomy_id = tt.term_taxonomy_id WHERE tt.taxonomy IN ('wp_theme') AND tr.object_id IN (518, 288) ORDER BY t.name ASC;
-SELECT user_id, meta_key, meta_value FROM wp_usermeta WHERE user_id IN (1) ORDER BY umeta_id ASC;
 SELECT * FROM wp_users WHERE ID IN (1);
 SELECT wp_posts.ID FROM wp_posts LEFT JOIN wp_term_relationships ON (wp_posts.ID = wp_term_relationships.object_id) WHERE 1=1 AND wp_posts.post_name IN ('header') AND ( wp_term_relationships.term_taxonomy_id IN (2) ) AND wp_posts.post_type = 'wp_template_part' AND ((wp_posts.post_status = 'publish')) GROUP BY wp_posts.ID ORDER BY wp_posts.post_date DESC LIMIT 0, 1;
 SELECT wp_posts.* FROM wp_posts WHERE ID IN (315);
```

spacedmonkey committed:
Merge branch 'trunk' into fix/lazy-loading-user-meta

spacedmonkey commented:
Committed in [[60989]](https://core.trac.wordpress.org/changeset/60989)
